### PR TITLE
Fix DS new assignment feature

### PR DESCRIPTION
**Pull request by Atlante45**
Fixes the "New Assignment" tab in the domain settings.

### CODE REVIEW

**Comment by ZappoMan** [comment]
CR looks good - if it passes tests, we can merge.

File: domain-server/src/DomainServer.cpp [hunk 1]
```diff
@@ -1152,13 +1152,22 @@ QJsonObject DomainServer::jsonObjectForNode(const SharedNodePointer& node) {
...
 
-const char ASSIGNMENT_SCRIPT_HOST_LOCATION[] = "resources/web/assignment";
+QDir pathForAssignmentScriptsDirectory() {
+    static const QString SCRIPTS_DIRECTORY_NAME = "/scripts/";
+
+    QDir directory(ServerPathUtils::getDataDirectory() + SCRIPTS_DIRECTORY_NAME);
+    if (!directory.exists()) {
+        directory.mkpath(".");
+        qInfo() << "Created path to " << directory.path();
+    }
```
Comment (birarda) [inline domain-server/src/DomainServer.cpp]: kind of a bummer to check this every time? Could we check if `pathForAssignmentScriptsDirectory()` exists on launch and create it if not?

**Comment by OmegaHeron** [comment]
Building source branch to test on Linux now

**Comment by hifi-gustavo** [comment]
Build console output for commit: edd9432eb46d43ba000f0b3799db1bbd7a85b45a available at: https://gist.github.com/f4c3910f4a282136dd09<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.dmg

**Comment by OmegaHeron** [comment]
This *seems* to work fine on Linux - it's worth saying I didn't have an existing persistent script setup so it works cleanly starting from nothing.  Mileage may vary on a setup having scripts in place under old code.

**Comment by ZappoMan** [comment]
I don't know... I just tested the windows PR build, and it doesn't seem to work... this is what my Assignment client logs say:
```
[01/21 21:03:58] [DEBUG] [5448] [agent] Registering a packet listener for packet list type 57 (MessagesData)
[01/21 21:03:58] [DEBUG] [5448] [agent] Sending connect request to domain-server at "localhost"
[01/21 21:03:58] [DEBUG] [5448] [agent] Local domain-server port read from shared memory (or default) is 40102
[01/21 21:03:58] [DEBUG] [5448] [agent] AssetClient disk cache setup at "C:/Users/Brad/AppData/Local/High Fidelity - PR6893/assignment-client" (size: 10 GB)
[01/21 21:03:58] [DEBUG] [5448] [agent] Domain ID changed to "ceb78b61-023d-478c-a6ae-e4f80efdb01d"
[01/21 21:03:58] [DEBUG] [5448] [agent] NodeList UUID changed from "00000000-0000-0000-0000-000000000000" to "5de3d224-55aa-4dfd-af1c-8058c50c9552"
[01/21 21:03:58] [DEBUG] [5448] [agent] Added "Avatar Mixer" (W) {007a16b5-83a1-42dc-9503-133a0b61648e} 174.21.92.4:64788 / 192.168.0.11:64788 
[01/21 21:03:58] [DEBUG] [5448] [agent] Downloading script at "http://127.0.0.1:40100/assignment/5de3d224-55aa-4dfd-af1c-8058c50c9552/"
[01/21 21:03:58] [DEBUG] [5448] [agent] Added "Entity Server" (o) {e49bf502-7826-4781-88bf-f73169053b8c} 174.21.92.4:60815 / 192.168.0.11:60815 
[01/21 21:03:58] [DEBUG] [5448] [agent] Added "Audio Mixer" (M) {d36ba8d8-95c3-42dc-9d07-cf44c8bf4ed2} 174.21.92.4:64789 / 192.168.0.11:64789 
[01/21 21:03:58] [DEBUG] [5448] [agent] Added "Asset Server" (A) {06df1d22-2235-499f-a868-6bf1e851819b} 174.21.92.4:60818 / 192.168.0.11:60818 
[01/21 21:03:58] [DEBUG] [5448] [agent] Added "Messages Mixer" (m) {6d405378-532b-4c18-a757-2cf4e9f07898} 174.21.92.4:60817 / 192.168.0.11:60817 
[01/21 21:03:58] [DEBUG] [5448] [agent] Activating local socket for network peer with ID "007a16b5-83a1-42dc-9503-133a0b61648e"
[01/21 21:03:58] [DEBUG] [5448] [agent] Activating local socket for network peer with ID "e49bf502-7826-4781-88bf-f73169053b8c"
[01/21 21:03:58] [DEBUG] [5448] [agent] Activating local socket for network peer with ID "d36ba8d8-95c3-42dc-9d07-cf44c8bf4ed2"
[01/21 21:03:58] [DEBUG] [5448] [agent] Activating local socket for network peer with ID "06df1d22-2235-499f-a868-6bf1e851819b"
[01/21 21:03:58] [DEBUG] [5448] [agent] Activating local socket for network peer with ID "6d405378-532b-4c18-a757-2cf4e9f07898"
[01/21 21:03:58] [DEBUG] [5448] [agent] Downloaded script: "HTTP/1.1 404 Not Found\r\nContent-Length: 19\r\nContent-Type: text/plain; charset=ISO-8859-1\r\nConnection: close\r\n\r\nResource not found."
```

**Comment by ZappoMan** [comment]
@OmegaHeron - how did you test this? The repro we should test is go to the Domain server page, choose the Assignments > New Assignment menu... and then write (or copy/paste) a script and click Run... the AC should launch and run the script.

**Comment by OmegaHeron** [comment]
@ZappoMan - retesting and will report back.  I had to compile and setup a isolated instance to try this on linux.  It may well be I didn't actually pull/build what I thought I did/wanted to pull and built the unmodified version.  I'll test on Windows via localhost just to speed things up.  Getting late here.

**Comment by OmegaHeron** [comment]
On a windows 10 box that stack had never ran on - I installed this PR - added a task by pasting in some js - clicked run - no error reported and I find in;

`C:\Users\jl\AppData\Roaming\High Fidelity - PR6893\domain-server\scripts`
All the scripts I added and engaged (as uuid named files).  It may be this only works on a machine that's not ran stack before?  Tried again on linux and behaved much the same.

**Comment by ZappoMan** [comment]
hmmm - does your script do something obvious? what script did you use? I
tried something that would print to the log... so I expect it to write to
the agent log.. but it doesn't

**Comment by ZappoMan** [comment]
Ok, I just built this branch locally and ran it, and I get the same behavior as the PR build...

If I run a command line assignment client type 2, I get the following...

```
[01/21 21:49:33] [DEBUG] [9632] [agent] Downloading script at "http://127.0.0.1:40100/assignment/8796a50f-6f31-4391-86c4-880b8b752b2f/"
[01/21 21:49:33] [DEBUG] [9632] [agent] Downloaded script: "HTTP/1.1 404 Not Found\r\nContent-Length: 19\r\nContent-Type: text/plain; charset=ISO-8859-1\r\nConnection: close\r\n\r\nResource not found."
[01/21 21:49:33] [DEBUG] [9632] [agent] Changing face model for avatar to ""
[01/21 21:49:33] [DEBUG] [9632] [agent] Changing skeleton model for avatar to "file:///c:/Development/HiFi/testing/Atlante45/build/assignment-client/Release/resources/meshes/defaultAvatar_full.fst"
[01/21 21:49:33] [DEBUG] [9632] [agent] Registering a packet listener for packet list type 11 (BulkAvatarData)
[01/21 21:49:33] [DEBUG] [9632] [agent] Registering a packet listener for packet list type 5 (KillAvatar)
[01/21 21:49:33] [DEBUG] [9632] [agent] Registering a packet listener for packet list type 29 (AvatarIdentity)
[01/21 21:49:33] [DEBUG] [9632] [agent] Registering a packet listener for packet list type 30 (AvatarBillboard)
[01/21 21:49:33] [WARNING] [9632] [agent] DependencyManager::get(): No instance available for class controller::ScriptingInterface
[01/21 21:49:33] [WARNING] [9632] [agent] [SyntaxError] Expected `;', `;' in :1(10)
```

**Comment by OmegaHeron** [comment]
If you want to paste in what you're running I'll give it a go here - I just added a do nothing thing looping and looked for task to show up in DS Node display and checked log for anything off.  I'm digging through log now to see if something got past me.

**Comment by ZappoMan** [comment]
@OmegaHeron - I am pretty certain your script isn't actually running... unless this is a Windows 8 OS specific problem... which I guess it could be...

**Comment by ZappoMan** [comment]
Also note: as Omega also notes -- there is an Agent node sitting in my Node list on my domain page... and the agent is sitting there still in the "Agent" mode, it hasn't dropped back to a waiting assignment-client.

**Comment by OmegaHeron** [comment]
I'm suspecting now that it's not.  It tricked me into thinking it is expletive deleted.  It's asking for an assignment type **7** when I try to run a script printing "Hello World!" to console...I thought script was 2?

**Comment by ZappoMan** [comment]
Here's the script I tried...

```
print("yo");
Script.stop();
```

That should basically print "yo" to the log, and then stop the script running and become a free ac again.

**Comment by OmegaHeron** [comment]
Right - I didn't add stop thus it persists with it showing an "extra" agent in node list.  Server Console doesn't alloc any spare servers for running scripts?  I'll manually start a type 2 and see what I see.

**Comment by OmegaHeron** [comment]
Using your two liner I see;
`[01/22 01:03:03] [DEBUG] Saved a script for assignment at C:/Users/jl/AppData/Roaming/High Fidelity - PR6893/domain-server/scripts/072d2a57-3b92-4d77-96f2-2d3d279bce1c`
`[01/22 01:03:03] [DEBUG] Deploying assignment - UUID: {072d2a57-3b92-4d77-96f2-2d3d279bce1c}, Type: 2 - to 127.0.0.1:58370`
`[01/22 01:03:03] [DEBUG] Assignment deployed with "707b0954-ddaa-4424-aa5f-44913dcb2447" matches unfulfilled assignment "072d2a57-3b92-4d77-96f2-2d3d279bce1c"`
`[01/22 01:03:03] [DEBUG] Added "Agent" (I) {fe3c9ca5-558d-46a9-bf64-3a5984744bf6} 70.127.20.117:58370 / 192.168.0.4:58370 `
`[01/22 01:03:03] [DEBUG] Activating symmetric socket for network peer with ID "fe3c9ca5-558d-46a9-bf64-3a5984744bf6"`

No print of yo to console.

**Comment by OmegaHeron** [comment]
Without my AV logged in it's able to (attempt) run so I guess server console isn't spooling up any spares and the type 7 was a red herring covering it not being able to find a spare to morph to a type 2.

**Comment by OmegaHeron** [comment]
Ok - finally - after restarting about 5 times - got this in log (while running spare AC in command window).  It starts up normal - first time I press run in script editor nothing happened - pressed again and we're off.  It throws 404 error trying to load script... 08084539-fdf8-4aa8-90ea-dadb2cc1b333

Looking in AppData I find script 08084539-fdf8-4aa8-90ea-dadb2cc1b333 with contents;

// Here you can put a script that will be run by an assignment-client (AC)
// For examples, please go to http://public.highfidelity.io/scripts
// The directory named acScripts contains assignment-client specific scripts you can try.
print("yo");
Script.stop();

Consider my premature - it's working - withdrawn.

[01/22 01:17:10] [DEBUG] [1696] [assignment-client] Received an assignment - UUID: {66fe7d2d-0dc8-4447-8c06-4626399424bb}, Type: 2
[01/22 01:17:10] [DEBUG] [1696] [assignment-client] Resetting current domain connection information.
[01/22 01:17:10] [DEBUG] [1696] [assignment-client] Hard reset in NodeList DomainHandler.
[01/22 01:17:10] [DEBUG] [1696] [assignment-client] Destination IP for assignment is "127.0.0.1"
[01/22 01:17:10] [DEBUG] [1696] [agent] Sending connect request to domain-server at "localhost"
[01/22 01:17:10] [DEBUG] [1696] [agent] Registering a packet listener for packet list type 57 (MessagesData)
[01/22 01:17:10] [DEBUG] [1696] [agent] Local domain-server port read from shared memory (or default) is 40102
[01/22 01:17:10] [DEBUG] [1696] [agent] Domain ID changed to "eb59c3f4-702b-4d69-91c1-a02c62ff7f04"
[01/22 01:17:10] [DEBUG] [1696] [agent] NodeList UUID changed from "00000000-0000-0000-0000-000000000000" to "87117b32-3848-4f38-86a6-8746a2a07058"
**[01/22 01:17:10] [DEBUG] [1696] [agent] Downloading script at "http://127.0.0.1:40100/assignment/87117b32-3848-4f38-86a6-8746a2a07058/"**
[01/22 01:17:10] [DEBUG] [1696] [agent] Added "Audio Mixer" (M) {5c345f36-68a4-4245-9e4f-061fc3493a39} 70.127.20.117:58370 / 192.168.0.4:58370
[01/22 01:17:10] [DEBUG] [1696] [agent] Added "Messages Mixer" (m) {58f2e8e3-2984-4094-a23b-0a48add395d3} 70.127.20.117:55234 / 192.168.0.4:55234
**[01/22 01:17:10] [DEBUG] [1696] [agent] Downloaded script: "HTTP/1.1 404 Not Found\r\nContent-Length: 19\r\nContent-Type: text/plain; charset=ISO-8859-1\r\nConnection: close\r\n\r\nResource not found."**

File: domain-server/src/DomainServer.cpp [hunk 2]
```diff
@@ -1203,13 +1211,15 @@ bool DomainServer::handleHTTPRequest(HTTPConnection* connection, const QUrl& url
...
-            // have the HTTPManager serve the appropriate script file
-            return _httpManager.handleHTTPRequest(connection, scriptURL, true);
+            QFile scriptFile(pathForAssignmentScript(matchingAssignment->getUUID()));
+
+            if (scriptFile.exists() && scriptFile.open(QIODevice::ReadOnly)) {
+                connection->respond(HTTPConnection::StatusCode200, scriptFile.readAll(), "application/javascript");
+                return true;
+            } else {
+                connection->respond(HTTPConnection::StatusCode404, "Resource not found.");
+                return false;
```
Comment (birarda) [inline domain-server/src/DomainServer.cpp]: Double check, but I think this return means you're saying you did not handle the request. You did by serving the 404.

Comment (Atlante45) [inline domain-server/src/DomainServer.cpp]: Oh right, fixed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 612311afb477c2dc8a23cde653285665b538b1bb available at: https://gist.github.com/f4c3910f4a282136dd09<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9330c3d4f32162f7f8d247d1217dcaee75701f6b available at: https://gist.github.com/f4c3910f4a282136dd09<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6893.dmg

**Comment by OmegaHeron** [comment]
Tested on Windows 10 - Ran script;
print("Yo!");
Script.Stop();

Logged output:
`[01/22 19:23:38] [DEBUG] [8364] [agent] script:print()<< Yo!`

**Comment by OmegaHeron** [comment]
Tested OS X 10.11.1 works exactly as example given above for windows.  For this simple test case it passes on both systems.

**Pull request merged**